File: memory-management/MemoryManagement.java
import java.util.ArrayList;
import java.util.List;

class BigObject {
    int[] data;

    BigObject(int size) {
        this.data = new int[size];
    }
}

public class MemoryManagement {

    static void lifetimeDemo() {
        System.out.println("\n-- Object lifetime --");
        BigObject obj = new BigObject(100_000);
        System.out.println("Object created, len(data)= " + obj.data.length);
        obj = null;
        System.out.println("Reference cleared; object is eligible for GC");
    }

    static void gcDemo() {
        System.out.println("\n-- Garbage collection --");
        Runtime rt = Runtime.getRuntime();
        long before = rt.totalMemory() - rt.freeMemory();
        System.out.println("Used memory before: " + before / (1024 * 1024) + " MB");

        List<BigObject> tmp = new ArrayList<>();
        for (int i = 0; i < 200; i++) {
            tmp.add(new BigObject(20_000));
        }
        tmp = null;

        System.gc();
        long after = rt.totalMemory() - rt.freeMemory();
        System.out.println("Used memory after:  " + after / (1024 * 1024) + " MB");
        System.out.println("Note: System.gc() is only a request to the JVM");
    }

    static void performanceTradeoff() {
        System.out.println("\n-- Performance tradeoff --");
        long start = System.nanoTime();
        List<Object> list = new ArrayList<>();
        for (int i = 0; i < 500_000; i++) {
            list.add(new Object());
        }
        long elapsedMs = (System.nanoTime() - start) / 1_000_000;
        System.out.println("Allocated 500000 objects in " + elapsedMs + " ms");
        System.out.println("Tradeoff: GC reduces leaks/dangling pointers but adds runtime overhead");
    }

    public static void main(String[] args) {
        System.out.println("Language: Java");
        lifetimeDemo();
        gcDemo();
        performanceTradeoff();
    }
}

File: memory-management/memory_management.py
import gc
import time


class BigObject:
    def __init__(self, size: int):
        self.data = [0] * size


def lifetime_demo() -> None:
    print("\n-- Object lifetime --")
    obj = BigObject(100_000)
    print("Object created, len(data)=", len(obj.data))
    del obj
    print("Reference deleted; Python uses ref counting + cyclic GC")


def gc_demo() -> None:
    print("\n-- Garbage collection --")
    print("GC counts before:", gc.get_count())
    tmp = [BigObject(10_000) for _ in range(100)]
    del tmp
    collected = gc.collect()
    print("Forced gc.collect() reclaimed objects:", collected)
    print("GC counts after:", gc.get_count())


def performance_tradeoff() -> None:
    print("\n-- Performance tradeoff --")
    start = time.perf_counter()
    nums = []
    for i in range(500_000):
        nums.append(i)
    elapsed = time.perf_counter() - start
    print(f"Allocated 500000 integers in {elapsed:.4f}s")
    print("Tradeoff: automatic memory management is safer but has GC/runtime overhead")


def main() -> None:
    print("Language: Python")
    lifetime_demo()
    gc_demo()
    performance_tradeoff()


if __name__ == "__main__":
    main()
